File: TouchNothing/Services/LeaderboardCache.swift
import Foundation

enum LeaderboardCache {
    static func saveBestSession(entries: [LeaderboardEntry]) {
        save(entries, forKey: StorageKeys.leaderboardCache)
    }

    static func loadBestSession() -> [LeaderboardEntry] {
        load(forKey: StorageKeys.leaderboardCache)
    }

    static func saveTotal(entries: [TotalLeaderboardEntry]) {
        save(entries, forKey: StorageKeys.totalLeaderboardCache)
    }

    static func loadTotal() -> [TotalLeaderboardEntry] {
        load(forKey: StorageKeys.totalLeaderboardCache)
    }

    static func clearAll() {
        UserDefaults.standard.removeObject(forKey: StorageKeys.leaderboardCache)
        UserDefaults.standard.removeObject(forKey: StorageKeys.totalLeaderboardCache)
    }

    private static func save<T: Encodable>(_ entries: [T], forKey key: String) {
        guard let encodedData = try? JSONEncoder().encode(entries) else { return }
        UserDefaults.standard.set(encodedData, forKey: key)
    }

    private static func load<T: Decodable>(forKey key: String) -> [T] {
        guard let cachedData = UserDefaults.standard.data(forKey: key) else {
            return []
        }
        return (try? JSONDecoder().decode([T].self, from: cachedData)) ?? []
    }
}

File: app/src/main/java/com/touchnothing/app/data/local/LeaderboardCache.kt
package com.touchnothing.app.data.local

import android.content.Context
import com.touchnothing.app.data.model.LeaderboardEntry
import com.touchnothing.app.data.model.TotalLeaderboardEntry
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

class LeaderboardCache(
    private val prefs: android.content.SharedPreferences,
    private val json: Json,
) {
    constructor(context: Context) : this(
        context.getSharedPreferences(LocalUserStore.PREFS_NAME, Context.MODE_PRIVATE),
        Json { ignoreUnknownKeys = true },
    )

    fun saveBestSession(entries: List<LeaderboardEntry>) {
        save(entries, StorageKeys.LEADERBOARD_CACHE)
    }

    fun loadBestSession(): List<LeaderboardEntry> = load(StorageKeys.LEADERBOARD_CACHE)

    fun saveTotal(entries: List<TotalLeaderboardEntry>) {
        save(entries, StorageKeys.TOTAL_LEADERBOARD_CACHE)
    }

    fun loadTotal(): List<TotalLeaderboardEntry> = load(StorageKeys.TOTAL_LEADERBOARD_CACHE)

    fun clearAll() {
        prefs.edit()
            .remove(StorageKeys.LEADERBOARD_CACHE)
            .remove(StorageKeys.TOTAL_LEADERBOARD_CACHE)
            .apply()
    }

    private inline fun <reified T> save(entries: List<T>, key: String) {
        prefs.edit().putString(key, json.encodeToString(entries)).apply()
    }

    private inline fun <reified T> load(key: String): List<T> {
        val cached = prefs.getString(key, null) ?: return emptyList()
        return runCatching { json.decodeFromString<List<T>>(cached) }
            .getOrDefault(emptyList())
    }
}
